File: src/resolver/walk.ts
import { promises as fs } from "node:fs";
import * as path from "node:path";

/**
 * Recursively enumerate every regular file under `dir`, returning entries
 * relative to `dir` in lex-sorted, posix-style form.
 *
 * Symlink behavior (v1):
 *  - Symlinks-to-files are followed and returned as files.
 *  - Symlinks-to-directories *inside* a contributor's `.claude/` subtree are
 *    NOT traversed (no cycle protection, so this avoids accidental loops).
 *  - The contributor root itself may still be a symlink to a directory —
 *    callers resolve existence via `isDirectory` (uses `fs.stat`) before
 *    invoking this function, so a symlinked component dir works at the root.
 *
 * Returns [] if `dir` does not exist or is not a directory; callers higher
 * up are responsible for validating existence (R7).
 */
export async function walkClaudeDir(
  dir: string,
): Promise<Array<{ relPath: string; absPath: string }>> {
  let stat: Awaited<ReturnType<typeof fs.stat>>;
  try {
    stat = await fs.stat(dir);
  } catch (err) {
    const e = err as NodeJS.ErrnoException;
    if (e.code === "ENOENT") return [];
    throw err;
  }
  if (!stat.isDirectory()) return [];

  const out: Array<{ relPath: string; absPath: string }> = [];
  await walk(dir, "", out);
  out.sort((a, b) => (a.relPath < b.relPath ? -1 : a.relPath > b.relPath ? 1 : 0));
  return out;
}

async function walk(
  base: string,
  rel: string,
  out: Array<{ relPath: string; absPath: string }>,
): Promise<void> {
  const here = rel === "" ? base : path.join(base, rel);
  const entries = await fs.readdir(here, { withFileTypes: true });
  for (const entry of entries) {
    const childRel = rel === "" ? entry.name : `${rel}/${entry.name}`;
    const childAbs = path.join(here, entry.name);
    if (entry.isDirectory()) {
      await walk(base, childRel, out);
    } else if (entry.isFile() || entry.isSymbolicLink()) {
      // Follow symlink-to-file by stat'ing through; if stat fails, treat as
      // if absent (best-effort; the materializer in E3 will hard-fail later).
      if (entry.isSymbolicLink()) {
        try {
          const targetStat = await fs.stat(childAbs);
          if (!targetStat.isFile()) continue;
        } catch {
          continue;
        }
      }
      out.push({ relPath: childRel, absPath: childAbs });
    }
  }
}

export async function isDirectory(p: string): Promise<boolean> {
  try {
    const s = await fs.stat(p);
    return s.isDirectory();
  } catch {
    return false;
  }
}

File: internal/resolver/walk.go
package resolver

import (
	"errors"
	"io/fs"
	"os"
	"path"
	"path/filepath"
	"sort"
)

// WalkEntry is a single file from walkClaudeDir / walkProfileRoot.
type WalkEntry struct {
	RelPath string
	AbsPath string
}

// WalkClaudeDir recursively enumerates every regular file under dir,
// returning entries relative to dir in lex-sorted, posix-style form.
//
// Symlink behavior (v1):
//   - Symlinks-to-files are followed and returned as files.
//   - Symlinks-to-directories inside a contributor's `.claude/` subtree are
//     NOT traversed (no cycle protection, so this avoids accidental loops).
//
// Returns nil if dir does not exist or is not a directory.
func WalkClaudeDir(dir string) ([]WalkEntry, error) {
	st, err := os.Stat(dir)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	if !st.IsDir() {
		return nil, nil
	}

	var out []WalkEntry
	if err := walkInto(dir, "", &out); err != nil {
		return nil, err
	}
	sort.Slice(out, func(i, j int) bool { return out[i].RelPath < out[j].RelPath })
	return out, nil
}

func walkInto(base, rel string, out *[]WalkEntry) error {
	here := base
	if rel != "" {
		here = filepath.Join(base, filepath.FromSlash(rel))
	}
	entries, err := os.ReadDir(here)
	if err != nil {
		return err
	}
	for _, entry := range entries {
		childRel := entry.Name()
		if rel != "" {
			childRel = path.Join(rel, entry.Name())
		}
		childAbs := filepath.Join(here, entry.Name())

		mode := entry.Type()
		if mode.IsDir() {
			if err := walkInto(base, childRel, out); err != nil {
				return err
			}
			continue
		}
		if mode.IsRegular() {
			*out = append(*out, WalkEntry{RelPath: childRel, AbsPath: childAbs})
			continue
		}
		if mode&os.ModeSymlink != 0 {
			// Symlink: stat through to determine target type. Do NOT recurse
			// into symlinked directories (cycle protection).
			target, err := os.Stat(childAbs)
			if err != nil {
				continue
			}
			if target.Mode().IsRegular() {
				*out = append(*out, WalkEntry{RelPath: childRel, AbsPath: childAbs})
			}
			continue
		}
	}
	return nil
}

// IsDirectory returns true iff p exists and is a directory.
func IsDirectory(p string) bool {
	s, err := os.Stat(p)
	if err != nil {
		return false
	}
	return s.IsDir()
}

// WalkProfileRoot discovers files at the profile root (peer of
// profile.json, sibling of `.claude/`) that materialize to the project
// root rather than under `.claude/`. Per cw6/§12 / R44–R46 the only such
// file in v1 is CLAUDE.md.
func WalkProfileRoot(profileDir string) ([]WalkEntry, error) {
	candidate := filepath.Join(profileDir, "CLAUDE.md")
	st, err := os.Stat(candidate)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil, nil
		}
		return nil, err
	}
	if !st.Mode().IsRegular() {
		return nil, nil
	}
	return []WalkEntry{{RelPath: "CLAUDE.md", AbsPath: candidate}}, nil
}
